controller/kubernetes_configmap_controller.go

```diff
@@ -0,0 +1,246 @@
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-kubernetes-configmap-controller"}),
+
+		cfmStoreSynced: configMapInformer.Informer().HasSynced,
+	}
+
+	configMapInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    kc.enqueueConfigMapChange,
+		UpdateFunc: func(old, cur interface{}) { kc.enqueueConfigMapChange(cur) },
+		DeleteFunc: kc.enqueueConfigMapChange,
+	})
+
+	return kc
+}
+
+func (kc *KubernetesConfigMapController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	kc.logger.Infof("Start")
+	defer kc.logger.Infof("Shutting down")
+
+	if !controller.WaitForCacheSync(kc.name, stopCh, kc.cfmStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesConfigMapController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesConfigMapController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesConfigMapController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		kc.logger.WithError(err).Warnf("Error syncing configmap %v", key)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	kc.logger.WithError(err).Warnf("Dropping configmap %v out of the queue", key)
+	kc.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (kc *KubernetesConfigMapController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", kc.name, key)
+	}()
+
+	namespace, cfmName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if err := kc.reconcileDefaultStorageClass(namespace, cfmName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (kc *KubernetesConfigMapController) reconcileDefaultStorageClass(namespace, cfmName string) error {
+	// The default storageclass configmap must be in the same namespace as the current KubernetesConfigMapController
+	if namespace != kc.namespace {
+		return nil
+	}
+
+	storageCFM, err := kc.ds.GetConfigMap(kc.namespace, types.DefaultStorageClassConfigMapName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			// The storageclass configmap was deleted therefore we delete the actual storageclass
+			err := kc.ds.DeleteStorageClass(types.DefaultStorageClassName)
+			if err != nil && !datastore.ErrorIsNotFound(err) {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	storageclass, err := buildStorageClassManifestFromConfigMap(storageCFM)
+	if err != nil {
+		return err
+	}
+
+	existingSC, err := kc.ds.GetStorageClassRO(types.DefaultStorageClassName)
```

Comment:
Is it possible the existing storage class has default value filled in but the one in the configmap doesn't, so we will constantly update the storageclass object in the system?

Reply:
Yeah, it is possible. 
I am thinking of 2 ways:
1. Autofill the default values for the storage class built from the config map. (This one is not future-proof  because Kubernetes may introduce new fields in the future)
1. Update the config map to contains the newly updated storage class. This method seems like a circle.

Reply:
I changed the logic so that we don't update the storage class by default unless there is a change in one of the nine field of the StorageClass struct. 

In the future, if Kubernetes introduces a new field to StorageClass, we can update the function `needToUpdateStorageClass` to check the new field.

Reply:
Is there any Kubernetes function to fill in the default values for StorageClass?

Reply:
@yasker 
I couldn't find a function that do that.

I found an interesting approach:  store a hash value of the ConfigMap's data field into an annotation of the storageclass.  If this hash value change -> we delete/create the storageclass.

Reply:
We don't need to store the hash value. We can store the whole config we want to apply in the annotation as a JSON string, e.g.

`storageclass.longhorn.io/applied: {xxxx}`

Reply:
Yeah, that will also work. But why not `hash`? :) Git use `hash` to comparing extensively

Reply:
yeah, hash works too. The only downside is hash doesn't show what's the applied version, but it should be fine in our case.